### ISSUE TITLE
Cannot convert timestamp to number back and forth

### ISSUE THREAD

**Issue by david-cortes** [numpy/numpy#14066]
**Steps to reproduce:**
* Create a `datetime64` object.
* Convert said object to its numeric representation.
* Try to convert it back to `datetime64`.

**Expected behavior:** should convert it back.

**Actual behavior:** throws error.

```python
import numpy as np
np.datetime64("2019-01-01").astype("datetime64[s]").astype(int)
>>> 1546300800
np.datetime64(1546300800, "s")
>>> numpy.datetime64('2019-01-01T00:00:00')
np.datetime64(np.datetime64("2019-01-01").astype("datetime64[s]").astype(int), "s")
```
```
ValueError: Could not convert object to NumPy datetime
```

**Comment by seberg** [comment]
Yes, the scalar conversion code currently only accepts python integers. I am frankly not sure if we should relax it, since the whole integer conversion seems a bit unsafe to me? But it may be worth thinking about it.

Note that `np.array` and `astype()` should be able to convert it (using unsafe casting, but that is the default).

**Comment by anntzer** [comment]
At least `np.datetime64(np.int64(1), "s")` should succeed (it currently fails, this is basically the example above) as the casting is not unsafe (the underlying integer type is int64...).
I would argue that even the "unsafe" cases should work, as numpy doesn't complain on either `np.int8(np.int64(123456))` or `np.int64(np.int8(123))`.

**Comment by seberg** [comment]
@anntzer true, but for `np.int8(np.int64(123456))`, I do think that it may be best to aslo give an error. To be honest, I think the "correct" way forward (not something to happen soon I guess), would be to rather have `np.datetime.from_epoch()` or so to give you a way to do the unsafe thing. Or we add a casting parameter to the coercion.

The casting in your `np.datetime64(np.int64(1), "s")` example is unsafe in my book. It cannot fail maybe but if you look at it from a type safety perspective a datetime and an integer are very different beasts.

**Comment by anntzer** [comment]
If you think that `np.datetime64(np.int64(1), "s")` should raise, do you think `np.datetime64(1, "s")` should also raise?  If anything the latter is more unsafe, as it can involve truncation (if the value is large enough).

**Comment by seberg** [comment]
Actually the latter will raise if the value is too large. In any case, that would be my ideal choice (i.e. deprecate it). Just because it fuzzes type safety boundaries. I have to admit I do not work too much with datetimes, but I do not see why a `from_epoch()` or so classmethod would not be a better candidate. That said, we would have to add such a classmethod first before deprecating anything of course, and I am not sure how involved that is.

**Comment by anntzer** [comment]
Ah, fair point re: raising.
I'm fine with from_epoch too.  I don't work much with datetimes either, but in the rare cases where I do, their behavior not matching casting between integers of different widths (int8/int64, which you said should error out too) is a bit jarring.

**Comment by david-cortes** [comment]
I think `from_epoch` could be a potential solution, but it's certainly not intuitive. The rest of the types are all convertible as `np.type(input)` (same as it works in other languages), and at least for me, that's the first thing I'd try out of my head, and wouldn't like to find out later that it works fine in some situations but not in others.

**Comment by seberg** [comment]
You can raise an error which tells you "use `np.datetime64.fromtimestamp()`". For the rest of the types, they mostly mirror python (i.e. python allows coercion from float to int, and actually they are backing down from that now). Python does not allow coercion from int to datetime though.
